pyexec: Add event-driven variant pyexec_friendly_repl().

pfalcon opened the pull request:
pyexec_friendly_repl_process_char() and friends.

Unlike readline(), this was implemented in less formal, trial&error process.
So, original loop-based pyexec_friendly_repl() is left intact so far.

pfalcon commented:
Message says it all - I had to hack around trying to recover editing capabilities. I believe all the basic editing now works, though that needs to be tested more. But there're bunch of "extra" features not supported, like restarting repl ("soft" reset), switching repl modes, hard reset.

I'm again looking for suggestions how to proceed further, taking into account that this work was started to have progress with esp8266 port, not to start long refactors. My guess is that realistically it doesn't make sense to block it waiting for complete refactor (with all features) of friendly repl, but rather have both usual and event-driven version, selectable by config option, and leave further refactoring for later.

dpgeorge commented:
Yeah, it's not very elegant :)  Well, just merge it and we won't use this function on stmhal.  Keep original code; no need for config option, since if the function's not used it won't be linked in to the final binary.

dpgeorge commented on stmhal/pyexec.c:
```diff
@@ -195,6 +195,112 @@ int pyexec_raw_repl(void) {
+        return 0;
+    }
+
+    return 0;
```
Not needed?

pfalcon replied:
Last "return 0" you mean? gcc complains.

dpgeorge replied:
Really? But the if/else covers all cases.

pfalcon replied:
... and goto confuses gcc. I'll recheck.

pfalcon replied:
Well, issue is gone now, apparently I had goto outside if.

dpgeorge commented on stmhal/pyexec.h:
```diff
@@ -32,9 +32,12 @@ typedef enum {
 int pyexec_raw_repl(void);
 int pyexec_friendly_repl(void);
 int pyexec_file(const char *filename);
+void pyexec_friendly_repl_init();
```
(void)

pfalcon commented:
> no need for config option, since if the function's not used it won't be linked in to the final binary.

That's good, but a bit too far on "pragmatic" way ;-). I'd prefer to have such option at least for basic documentation purposes.

pfalcon commented:
With #1062 having been merged, this pullreq can now be tested in comfort on a hosts with:
~~~~
diff --git a/bare-arm/main.c b/bare-arm/main.c
index b56af43..1a9c3fa 100644
--- a/minimal/main.c
+++ b/minimal/main.c
@@ -10,6 +10,7 @@
 #include "py/pfenv.h"
 #include "py/gc.h"
 #include "pyexec.h"
+#include "pybstdio.h"
 
 void do_str(const char *src) {
     mp_lexer_t *lex = mp_lexer_new_from_str_len(MP_QSTR__lt_stdin_gt_, src, strlen(src), 0);
@@ -59,7 +60,13 @@ int main(int argc, char **argv) {
     gc_init(heap, heap + sizeof(heap));
 #endif
     mp_init();
-    pyexec_friendly_repl();
+    pyexec_friendly_repl_init();
+    for (;;) {
+        int c = stdin_rx_chr();
+        if (pyexec_friendly_repl_process_char(c)) {
+            break;
+        }
+    }
     //do_str("print('hello world!', list(x+1 for x in range(10)), end='eol\\n')");
     mp_deinit();
     return 0;

~~~~

pfalcon commented:
> Yeah, it's not very elegant :)

Yep, but I thought we don't what to pull protothreads in just for these few low-level functions ;-). Btw, with recent C++ talk, C++ is going to have efficient coroutine support https://www.youtube.com/watch?v=KUhSjfSbINE , http://blogs.msdn.com/b/vcblog/archive/2014/11/12/resumable-functions-in-c.aspx

Other comments, I've added readline_note_newline() to mp-readline API to accommodate repl needs. This just makes explicit an action which previously was implicit, to blindly follow client-driven flow (i.e. previously, readline() returned with complete input line, as confirmed by user; if it was called again with non-empty string it (apparently) assumed that all existing content in that string are previous lines of multi-line input; readline_note_newline() implements explicit action of "from now on, start to assume that any existing content in previously configured line buffer belongs to previous lines of multiline"). This is probably not ideal, and this stuff can be done implicitly on readline_process_char() side - I just don't know enough of readline() design/requirements and would need help with refactoring this. Not necessarily immediate, but then optimizing it should rather be postponed until it happens.

pfalcon commented:
> https://www.youtube.com/watch?v=KUhSjfSbINE , http://blogs.msdn.com/b/vcblog/archive/2014/11/12/resumable-functions-in-c.aspx

And here's more formal definition of "stackless" (https://github.com/CppCon/CppCon2014/blob/master/Presentations/await%202.0%20-%20Stackless%20Resumable%20Functions/await%202.0%20-%20Stackless%20Resumable%20Functions%20-%20Gor%20Nishanov%20-%20CppCon%202014.pdf):

> Stackful / Stackless
• How much state coroutine has? Just the locals of the coroutine or entire stack?

Commit:
pyexec: Add event-driven variant pyexec_friendly_repl().

pyexec_friendly_repl_process_char() and friends.

Unlike readline(), this was implemented in less formal, trial&error process.
So, original loop-based pyexec_friendly_repl() is left intact so far.

pfalcon commented:
Updated PR.

dpgeorge commented:
> Btw, with recent C++ talk, C++ is going to have efficient coroutine support

Great.  So in 10 years time we can rely on it being properly supported in majority of embedded targets and start using it.  And by then the language is so complex that it takes 1 week to compile our template-only code :)

pfalcon commented:
> Great. So in 10 years time we can rely on it being properly supported in majority of embedded targets

Well, it all accelerates lately. For example, Microsoft can't get right (according to @stinos) 15-year old standard for the language 80% is stuff still being written in, but look, they lead with this C++ coroutine implementation ;-). Besides, to support it for any embedded target, it will take to only write codegenerator for LLVM IR. Should be easy ;-).

Anyway, the patch is ready from my side, let me know if you'd like to test it more, or I just should proceed with it and integrating into esp8266 port.

dpgeorge commented:
> Besides, to support it for any embedded target, it will take to only write codegenerator for LLVM IR. Should be easy ;-).

:)

> Anyway, the patch is ready from my side, let me know if you'd like to test it more, or I just should proceed with it and integrating into esp8266 port.

Yes, you can go ahead.

Only thing I have to say is re the config variable MICROPY_EVENT_DRIVEN_REPL.  It would be nice to have some kind of systematic naming scheme for things like this, since we will be growing these config options without bound for some time.  I suggest putting in rough namespace, eg in this case by putting the word "REPL" first: MICROPY_REPL_EVENT_DRIVEN.  So "REPL" is the namespace.

pfalcon commented:
Sounds good. Didn't try to follow that route because we have MICROPY_HELPER_REPL

pfalcon commented:
Merged. Can be easily tested with d7337f288e9411d8473c6c77aa35a00b50346ca7.

dpgeorge commented:
> Didn't try to follow that route because we have MICROPY_HELPER_REPL

That's because that config option is in the "HELPER" namespace.

pfalcon commented:
Are you sure it shouldn't be renamed REPL_HELPER?

dpgeorge commented:
At the time it made more sense to put the "helpers" together.  But now it seems better to make the namespace "RELP" and "LEXER" for the 2 existing helpers.
